### PR TITLE
Fixed count of filtered issues when api request.

### PR DESCRIPTION
**Pull request by hinoshiba**
The search results for "The issues" will respond with the total number of issues in repository instead of the number of search results.
This result in pages that do not have a value in `rel = last` on the response HTTP header "Link".

* The target API
	* `api/v1/repos/issues/search`
	* `api/v1/repos/{repo}/issues/`

If you create about 10 "The issues" and close 2 of them, the final link is `page=10` , even if `?limit=1&page=1&state=open`. In such cases, the final link should be `page=8`. This has been fixed.

And, there was an extra reservation for the slice the related code, so I fixed it.

### CODE REVIEW

**Review by 6543 — CHANGES_REQUESTED** [review]
https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR160-R171 and https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR179-R191 are same

https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR336-R342 and https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR350-R360 are quite similar

can you dedublicate it by store the Options Struct into a variable you can pass to both functions?

**Comment by hinoshiba** [comment]
> https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR160-R171 and https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR179-R191 are same
> 
> https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR336-R342 and https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR350-R360 are quite similar
> 
> can you dedublicate it by store the Options Struct into a variable you can pass to both functions?

Is possible. However, I don't think it makes much sense in terms of processing.
They are said to be the same, but they all differ in the content of "ListOptions", mainly "PageSize".

The "models.CountIssues" and "model.Issues" arguments are references.
I think either of these needs to be done before the reference can be used safely.
1. Not run in parallel.
2. Copy later.
I think it will probably run in parallel, so I didn't think it made much sense in terms of processing.

I think you can reuse the structure by splitting the option structure and using two arguments. However, because it reduces memory efficiency, we made redundant entries.

**Comment by 6543** [comment]
🤔

**Comment by 6543** [comment]
Will test it tomorow :)

**Comment by hinoshiba** [comment]
> > https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR160-R171 and https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR179-R191 are same
> > https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR336-R342 and https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR350-R360 are quite similar
> > can you dedublicate it by store the Options Struct into a variable you can pass to both functions?
> 
> Is possible. However, I don't think it makes much sense in terms of processing.
> They are said to be the same, but they all differ in the content of "ListOptions", mainly "PageSize".
> 
> The "models.CountIssues" and "model.Issues" arguments are references.
> I think either of these needs to be done before the reference can be used safely.
> 
> 1. Not run in parallel.
> 2. Copy later.
> 
> I think it will probably run in parallel, so I didn't think it made much sense in terms of processing.
> 
> I think I can reuse the structure by splitting the option structure and using two arguments. However, because it reduces memory efficiency, I made redundant entries.

I missed.
I didn't actually look back and forth. There is no parallel processing. ;-D
Sorry,

Will update it after confirming the test.

File: routers/api/v1/repo/issue.go [hunk 1]
```diff
@@ -165,14 +170,32 @@ func SearchIssues(ctx *context.APIContext) {
...
+			IsPull:             isPull,
+		})
+		if err != nil {
+			ctx.Error(http.StatusInternalServerError, "CountIssues", err)
+			return
+		}
 	}
 
-	ctx.SetLinkHeader(issueCount, setting.UI.IssuePagingNum)
+	ctx.SetLinkHeader(int(filteredCount), setting.UI.IssuePagingNum)
```
Comment (6543) [inline routers/api/v1/repo/issue.go:177]:
```suggestion
	ctx.SetLinkHeader(int(filteredCount), setting.UI.IssuePagingNum)
	ctx.Header().Set("X-Total-Count", fmt.Sprintf("%d", filteredCount))
```

File: routers/api/v1/repo/issue.go [hunk 2]
```diff
@@ -165,14 +170,32 @@ func SearchIssues(ctx *context.APIContext) {
...
+			return
+		}
 
-	if err != nil {
-		ctx.Error(http.StatusInternalServerError, "Issues", err)
-		return
+		filteredCount, err = models.CountIssues(&models.IssuesOptions{
+			ListOptions: models.ListOptions{
+				Page:     ctx.QueryInt("page"),
+				PageSize: issueCount,
```
Comment (6543) [inline routers/api/v1/repo/issue.go]:
```suggestion
				PageSize: setting.UI.IssuePagingNum,
```

Comment (6543) [reply routers/api/v1/repo/issue.go]: ☝️ if you dedub things this is not needed anyway ...

**Review by 6543 — CHANGES_REQUESTED** [review]

**Comment by 6543** [comment]
I would say dedub/refactor things and I'll check it again - for me it looks like there is something still missing but I'll dig deeper into it after you changed the code :)

File: routers/api/v1/repo/issue.go (hunk 2 above)
Comment (hinoshiba) [reply routers/api/v1/repo/issue.go]: This is for the purpose of getting the maximum value that matches the filter.
Therefore, the current maximum value is set as an option.

**Comment by 6543** [comment]
@hinoshiba let me know if you are ready for review again :)

**Comment by hinoshiba** [comment]
> @hinoshiba let me know if you are ready for review again :)

Ok, Just a moment.

**Comment by hinoshiba** [comment]
@6543 Finished update.

**Comment by 6543** [comment]
just format nits: https://github.com/hinoshiba/gitea/pull/1

File: routers/api/v1/repo/issue.go [hunk 3]
```diff
@@ -304,25 +322,33 @@ func ListIssues(ctx *context.APIContext) {
...
-		return
-	}
+		if issues, err = models.Issues(issuesOpt); err != nil {
+			ctx.Error(http.StatusInternalServerError, "Issues", err)
+			return
+		}
 
-	ctx.SetLinkHeader(ctx.Repo.Repository.NumIssues, listOptions.PageSize)
-	ctx.Header().Set("X-Total-Count", fmt.Sprintf("%d", ctx.Repo.Repository.NumIssues))
+		issuesOpt.ListOptions = models.ListOptions{
```
Comment (lafriks) [inline routers/api/v1/repo/issue.go:331]: Why is this needed?

File: routers/api/v1/repo/issue.go [hunk 4]
```diff
@@ -151,28 +155,37 @@ func SearchIssues(ctx *context.APIContext) {
...
 
-	if err != nil {
-		ctx.Error(http.StatusInternalServerError, "Issues", err)
-		return
+		if issues, err = models.Issues(issuesOpt); err != nil {
+			ctx.Error(http.StatusInternalServerError, "Issues", err)
+			return
+		}
+
+		issuesOpt.ListOptions = models.ListOptions{
```
Comment (lafriks) [inline routers/api/v1/repo/issue.go:168]: same here

File: routers/api/v1/repo/issue.go (hunk 3 above)
Comment (hinoshiba) [reply routers/api/v1/repo/issue.go:331]: The maximum value that matches the filter is needed to calculate the last page number.
So, It need a value from the zero to tail.

Comment (6543) [reply routers/api/v1/repo/issue.go:331]: There coud be a if statement in the setupSession for IssueListOptions when page==0 ignore limint but you have to look into the code

**Comment by 6543** [comment]
https://github.com/go-gitea/gitea/blob/0186497e676b531787045863647f08f076ed0b32/models/issue.go#L1104-L1113

so @hinoshiba some suggestions:

**Comment by 6543** [comment]
delete:
* https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR110-R117
* https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR96

File: routers/api/v1/repo/issue.go (hunk 3 above)
Comment (6543) [inline routers/api/v1/repo/issue.go:340]:
```suggestion
		issuesOpt.ListOptions = models.ListOptions{}
```

File: routers/api/v1/repo/issue.go [hunk 5]
```diff
@@ -304,25 +322,33 @@ func ListIssues(ctx *context.APIContext) {
...
-	}
+		if issues, err = models.Issues(issuesOpt); err != nil {
+			ctx.Error(http.StatusInternalServerError, "Issues", err)
+			return
+		}
 
-	ctx.SetLinkHeader(ctx.Repo.Repository.NumIssues, listOptions.PageSize)
-	ctx.Header().Set("X-Total-Count", fmt.Sprintf("%d", ctx.Repo.Repository.NumIssues))
+		issuesOpt.ListOptions = models.ListOptions{
+			Page:     0,
```
Comment (6543) [inline routers/api/v1/repo/issue.go:341]:
```suggestion
```

File: routers/api/v1/repo/issue.go [hunk 6]
```diff
@@ -304,25 +322,33 @@ func ListIssues(ctx *context.APIContext) {
...
+			PageSize: ctx.Repo.Repository.NumIssues,
```
Comment (6543) [inline routers/api/v1/repo/issue.go]:
```suggestion
```

File: routers/api/v1/repo/issue.go [hunk 7]
```diff
@@ -304,25 +322,33 @@ func ListIssues(ctx *context.APIContext) {
...
+		}
```
Comment (6543) [inline routers/api/v1/repo/issue.go:343]:
```suggestion
```

**Review by 6543 — CHANGES_REQUESTED** [review]

**Comment by 6543** [comment]
# Summary:

```diff
diff --git a/routers/api/v1/repo/issue.go b/routers/api/v1/repo/issue.go
index b5faa8235..f56834b26 100644
--- a/routers/api/v1/repo/issue.go
+++ b/routers/api/v1/repo/issue.go
@@ -93,7 +93,6 @@ func SearchIssues(ctx *context.APIContext) {
                opts.AllLimited = true
        }
 
-       issueCount := 0
        for page := 1; ; page++ {
                opts.Page = page
                repos, count, err := models.SearchRepositoryByName(opts)
@@ -107,14 +106,6 @@ func SearchIssues(ctx *context.APIContext) {
                }
                log.Trace("Processing next %d repos of %d", len(repos), count)
                for _, repo := range repos {
-                       switch isClosed {
-                       case util.OptionalBoolTrue:
-                               issueCount += repo.NumClosedIssues
-                       case util.OptionalBoolFalse:
-                               issueCount += repo.NumOpenIssues
-                       case util.OptionalBoolNone:
-                               issueCount += repo.NumIssues
-                       }
                        repoIDs = append(repoIDs, repo.ID)
                }
        }
@@ -174,10 +165,7 @@ func SearchIssues(ctx *context.APIContext) {
                        return
                }
 
-               issuesOpt.ListOptions = models.ListOptions{
-                       Page:     0,
-                       PageSize: issueCount,
-               }
+               issuesOpt.ListOptions = models.ListOptions{}
                if filteredCount, err = models.CountIssues(issuesOpt); err != nil {
                        ctx.Error(http.StatusInternalServerError, "CountIssues", err)
                        return
@@ -337,10 +325,7 @@ func ListIssues(ctx *context.APIContext) {
                        return
                }
 
-               issuesOpt.ListOptions = models.ListOptions{
-                       Page:     0,
-                       PageSize: ctx.Repo.Repository.NumIssues,
-               }
+               issuesOpt.ListOptions = models.ListOptions{}
                if filteredCount, err = models.CountIssues(issuesOpt); err != nil {
                        ctx.Error(http.StatusInternalServerError, "CountIssues", err)
                        return
```

**Comment by 6543** [comment]
or just asign `models.ListOptions{}` since PageSice must be greater than 0

**Review by hinoshiba — COMMENTED** [review]
> or just asign `models.ListOptions{}` since PageSice must be greater than 0

Ok, I also think that -1 is easier to understand. :-D

**Comment by 6543** [comment]
> delete:
> 
>     https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR110-R117
>     https://github.com/go-gitea/gitea/pull/12275/files#diff-6a82b880ff1ae1c2e9b577046e45a6cbR96

@hinoshiba dont forget to rm unused code

**Comment by 6543** [comment]
EDIT: thanks

**Comment by 6543** [comment]
`make fmt` is posible needed at this point ...

**Comment by hinoshiba** [comment]
> `make fmt` is posible needed at this point ...

Ok, I'm Update.

**Comment by 6543** [comment]
well github code suggestions have there up and downs ;)

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
can you resolve the conflict?

It looks like nobody had time to look at it :/

**Comment by hinoshiba** [comment]
Ok, I just updated.

File: routers/api/v1/repo/issue.go [hunk 8]
```diff
@@ -107,19 +106,12 @@ func SearchIssues(ctx *context.APIContext) {
 		}
 		log.Trace("Processing next %d repos of %d", len(repos), count)
 		for _, repo := range repos {
-			switch isClosed {
```
Comment (lunny) [inline routers/api/v1/repo/issue.go:110]: Why this is not right if it's consistent?

Comment (6543) [reply routers/api/v1/repo/issue.go:110]: consistent?

Comment (hinoshiba) [reply routers/api/v1/repo/issue.go:110]: Maybe I should answer the reason why I deleted is it?

The reason I removed it is because it is a counter that cannot include conditions "Issues Options".
I added function of counter that can include that rule at line 171.

Comment (zeripath) [reply routers/api/v1/repo/issue.go:110]: This PR doesn't need to use this coarse measure because it calculates the count exactly in models/issue.go:1264:func CountIssues(...) ...

File: models/issue.go [hunk 9]
```diff
@@ -1260,6 +1260,30 @@ func Issues(opts *IssuesOptions) ([]*Issue, error) {
...
+	defer sess.Close()
+
+	var cnt int64
+	countsSlice := make([]*struct {
+		RepoID int64
+		Count  int64
+	}, 0, len(opts.RepoIDs))
+
+	sess.GroupBy("issue.repo_id").
+		Select("issue.repo_id AS repo_id, COUNT(*) AS count").Table("issue")
```
Comment (zeripath) [inline models/issue.go]: Why the GroupBy? Why not just do simple Count?

Comment (hinoshiba) [reply models/issue.go]: Yes, I want to use function of simple count.
But It function is included bug when MSSQL.
So I reluctantly did use the GroupBy. -> https://github.com/go-gitea/gitea/pull/12275/commits/3f099650f75e7723b6d19616dea464d4e0840e85

Comment (zeripath) [reply models/issue.go]: Hmm that's interesting and looks like a potential bug in Xorm. However could you try:

```suggestion
	sess.Select("COUNT(*) AS count").Table("issue")
```

Or if that doesn't work:

```suggestion
	sess.Select("COUNT(issue.id) AS count").Table("issue")
```

Comment (hinoshiba) [reply models/issue.go]: It was try ;-D

Comment (hinoshiba) [reply models/issue.go]: And the same problem arose.

Comment (hinoshiba) [reply models/issue.go]: Sorry, This suggested code seems a bit different from what I tried.
I will check it tonight.

Comment (hinoshiba) [reply models/issue.go]: @zeripath 
thank you for your suggest code. I updated.

**Review by 6543 — APPROVED** [review]

**Review by zeripath — APPROVED** [review]
